### ISSUE TITLE
How to find out if a node is inside a finally block?

### ISSUE THREAD

**Issue by malteskoruppa** [javaparser/javaparser#1866]
I found it curiously difficult to determine, for a node, whether that node is inside a `finally` block. It feels like this is something that should be fairly easy, yet I ended up with rather complex code. Maybe I'm just overworked and ready for the week-end? :-)

Given a node `node` for which I want to find out whether it is inside a `finally` block, I have this:

```java
boolean inFinallyBlock = false;

// check whether the node is inside a try statement
Optional<TryStmt> tryParent = node.findAncestor(TryStmt.class);

// if so, check whether that try statement has a finally block and if that block contains our node
if (tryParent.isPresent()) {
	Optional<BlockStmt> finallyBlock = tryParent.get().getFinallyBlock();
	if (finallyBlock.isPresent()) {
		Optional<Node> sameNode = finallyBlock.get().findFirst(Node.class, n -> n == node);
		if (sameNode.isPresent()) {
			inFinallyBlock = true;
		}
	}
}
```

It's really lengthy and rather inefficient. I look upwards for a `try` node first, then downwards (from the `finally` block) to my original node again.

1. Am I just not seeing the wood for the trees? Is there a conceptually more elegant way to do this?

2. If not, how about a findAncestor(...) method that takes a predicate? Just like there are variants of findFirst(...) and findAll(...) with and without predicates, a findAncestor(...) with predicate could be useful to do the above much more elegantly. Instead of the above I could do _something like_:

    ```java
    node.findAncestor(BlockStmt.class, bs -> bs.getParent().isTryStmt() &&
                                             ((TryStmt) bs.getParent()).getFinallyBlock().isPresent() &&
                                             ((TryStmt) bs.getParent()).getFinallyBlock().get() == bs);
    ```

    Ok, it's not the embodiment of elegance, but at least it's shorter and more efficient. And one can probably massage this predicate a bit to make it more beautiful. :-)

**Comment by malteskoruppa** [comment]
I just noticed that the first variant which I wrote above (the one that first searches upwards, then downwards again) wouldn't even work properly in some contrived scenarios, e.g., when parsing code such as

```java
try {
  // ...
} catch (Exception e) {
  // ...
} finally {
  try {
    // ...
  } catch (Exception e) {
    foo(); // this is the node for which we want to find out whether it is inside a finally block; in this case, it is!
  }
}
```

The second variant (which would require a `findAncestor` method with predicate) would work, though.

**Comment by malteskoruppa** [comment]
Having rested a bit, I think I indeed did not see the most obvious solution: simply implement a findAncestor(...) method that takes a predicate myself. :grin: It's really a very, very easy-to-write method that simply iteratively traverses the parents and checks the given predicate on each of these parents (and returns on a match).

But nevertheless, it's a method that could be a useful addition to the JavaParser API, don't you think? Abstract Syntax Christmas Tree principle and all that. :-)

**Comment by malteskoruppa** [comment]
PS. With "implementing it myself", I meant "in my own project, as a helper method". But then I thought, why not just make it a pull request while I'm at it? :-)

**Comment by matozoid** [comment]
Thanks for the PR :-) The whole `find...` API is relatively new and I couldn't think of a reason to have that particular method back then. It seems you found a reason.

Another one that could be useful then would be the `findAllAncestors` one that returns a list, which is more general than both existing methods. But that's for another day.

**Comment by matozoid** [comment]
About the initial question: I guess it would have been easy if the AST had a `FinallyBlockStmt` or so, so that you could do a find ancestor on that. Not sure if it is worth adding though, and what other new nodes would have to be introduced, and what breakage would occur in the API.

**Comment by malteskoruppa** [comment]
> Another one that could be useful then would be the `findAllAncestors` one that returns a list, which is more general than both existing methods. But that's for another day.

Yes, I actually also thought about that too (but I didn't want to stray away from the issue at hand). I agree though: There are `findFirst` and `findAll` methods to find descendants, with and without predicates, respectively. So there should be `findAncestor` and `findAllAncestors` methods too, with and without predicates, respectively. That sounds like a nice consistent API. :-) (And one _might_ even consider renaming `findAncestor` *yet again* to `findFirstAncestor` for a consistent naming scheme.)

Another interesting question is this: IIRC the `findFirst` and `findAll` methods consider the start node itself too, which I think is a good thing. But then, shouldn't the `findAncestor` method (and a potential `findAllAncestors` method too) do the same (it currently doesn't)? Because...

- it'd be behaviorally consistent with `findFirst` / `findAll`;
- from an API point of view, I think it's nicer for users if the start node is considered by default. (It's easier to use a function that considers the start node and "filter it out" if one does not want to consider that start node, than it is to do the opposite: use a function that does not consider the start node and introduce boilerplate code to additionally consider the start node if one does want to consider that start node)

This then led me to the question: In graph theory, is a node usually considered an ancestor of itself? I am unsure if there is a universally agreed-upon definition. However, I tend to think that it simplifies a lot of things if definitions allow for nodes to be ancestors / descendants of themselves. Also, "Introduction to Algorithms" by CLRS (a well-established reference book on algorithms) seems to define it that way too, see [this question on SO](https://stackoverflow.com/questions/3078157/is-a-node-in-a-tree-considered-its-own-ancestor).

> About the initial question: I guess it would have been easy if the AST had a `FinallyBlockStmt` or so, so that you could do a find ancestor on that. Not sure if it is worth adding though, and what other new nodes would have to be introduced, and what breakage would occur in the API.

Yes, then it would have been easy. :-) I don't think it's worth introducing such extra nodes though. It feels like it'd unnecessarily complicate the AST structure. Using the predicates, there shouldn't be any worries now.

**Comment by matozoid** [comment]
Here, I'll tag this "improvement" so we can work on these methods someday.

About find/findAncestor: in my mind this whole API was about walking the tree. There are several algorithms, and I think breadth first is the default one. That means it includes the current node, since that's how these algorithms are defined. After that came walking in the other direction: up the tree through the parents. This is a strange thing since it is not about trees at all, it's more like list iteration. That's why most methods are just called "find", and one (now two) are called findAncestor.

But that's not an excuse to have a confusing API. Now, trying to find an ancestor and hitting `this` instead seems exactly the kind of confusion we want to avoid - I think we need to go full out @ftomassetti and add another `find...Descendants` API which excludes `this`... Damn, all of this makes me feel really uncomfortable :-/

**Comment by malteskoruppa** [comment]
I see, thanks for the explanations, it makes sense!

> But that's not an excuse to have a confusing API. Now, trying to find an ancestor and hitting `this` instead seems exactly the kind of confusion we want to avoid 

That's true too. I think it's fine if the `findAncestor` methods don't consider the start node. Maybe it's enough to simply stress these differences in the Javadocs, i.e., always include the information whether the start node is included or not in all of the `find*` methods. That way, users know for sure what to expect.

**Comment by matozoid** [comment]
One reason I'm feeling uncomfortable is because two users already stumbled because `this` was suddenly included while replacing a deprecated call... It's a pitfall, and it would be good to avoid those. Seeing how many questions are answered with "just look at the Javadoc" I would conclude that it does not always get read. Hm.

**Comment by malteskoruppa** [comment]
I'm also curious to see what @ftomassetti thinks! :-)

I think in this particular case, users stumbled upon this pitfall because the deprecated method suggested to replace the call with another method that did not exhibit exactly the same behavior. So that came as a surprise. But I don't think that it's a problem to consider the start node per se.

Also, of course an API should be designed in an intuitive way, but that does not justify ignoring the Javadoc. I think that as long as an API is sensible and the details are well-documented, everything's fine. Personally I wouldn't blame it on anyone but me if I made a mistake because I didn't look at the Javadoc. ;-)

**Comment by ftomassetti** [comment]
From my point of view I would say it would be nice if a node knew in which property of its parent its contained. I think we have some logic for this, see for example `findNodeListName` in `LexicalPreservingPrinter`.

Something like:

`if (node.getRole() == ObservableProperty.FINALLY_BLOCK) { ... }`

**Comment by ftomassetti** [comment]
where rolecould be implemented re-using some of the logic of `findNodeListName` or similarly using the metadata we have for node types. It would probably not have great performance though
